Detailed helptext framework

za419 opened the pull request:
Add the code required to generate and send a detailed helptext message, along with the config for the detailed helptext for the help command.

Usage:
![image](https://user-images.githubusercontent.com/15851480/103469270-511fc800-4d28-11eb-94ec-cb6c615a12ad.png)

za419 committed:
Add some config to support detailed helptext

I'm not writing all the helptext in this PR, but this config will help
demonstrate how to, and will guide implementation of the code to consume
it.

za419 committed:
Add functionality for unknown helptext request

za419 committed:
Add helptext generation

za419 committed:
Correct substring

za419 committed:
Fix typo

za419 committed:
Use sendLongMessage for detailed helptext

za419 committed:
Update title and subtitle formatting

za419 committed:
Indent list items further

(To make them more obvious)

za419 committed:
Remove second newline after title

za419 committed:
Blockquote detailed helptext

za419 committed:
Add a spare newline in lieu of a subtitle

JunhaoSLi commented on bot.js:
```diff
@@ -682,6 +683,75 @@ function command(message) {
+            }
+
+            // Now send the response.
+            sendLongMessage(message.channel, response);
```
Is sendLongReply something that could have been used here (but sendLongMessage was chosen instead)?

za419 replied:
sendLongReply could have been used as well here without issues, yes (`sendLongReply(message, response)`)

I chose sendLongMessage instead because I disliked how @za419, <title> looked... I suppose I could have added a new line at the beginning instead.

But overall I think sendLongMessage is appropriate here because helptext isn't necessarily person-specific - There's no reason it has to be directed at the sending user.

Of course, then you'd ask, why does short helptext use reply?
... And the answer is it was one of the first commands I wrote, and I didn't know how to send a message in the same channel without replying, and never bothered fixing it. I suppose they should be consistent though...

za419 committed:
Add example %cmd references to helptext

Jun sold me on my original suggestion for this syntax.

za419 committed:
Handle %cmd syntax

za419 committed:
Put examples in code blocks

za419 commented on default-config.json:
```diff
@@ -33,6 +33,17 @@
+      "subtitle": "The help command provides information to assist with the usage of Cadence.",
+      "lines": [
+        "It can be used in one of two ways:",
+        "    1. Without any argument (`%help`), it will generate a summary of commands and a brief snippet of their purpose.",
```
After my offline conversation with @JunhaoSLi, we agreed this would be a good syntax to insert the value of `config.commands.help` - IE the triggering phrase for the specified command. Here's how this looks:

![image](https://user-images.githubusercontent.com/15851480/103470391-72d47b80-4d37-11eb-8a7e-29cf3b97d983.png)

za419 commented on bot.js:
```diff
@@ -682,6 +683,116 @@ function command(message) {
+
+                                // If the first word is a command's internal name...
+                                if (
+                                    Object.keys(config.commands).includes(
```
This is one of the cases where I am *really* not a fan of Prettier's chosen formatting...

za419 commented on bot.js:
```diff
@@ -682,6 +683,116 @@ function command(message) {
+                                    )
+                                ) {
+                                    // Then add the command's trigger phrase to result
+                                    result += config.commands[firstWord];
```
Question - Should this be trimmed?

The trigger phrase for search is `Cadence search ` - With the space at the end. In other words, the message "Cadence searchrailgun" won't be noticed as anything by CadenceBot, but "Cadence search railgun" will trigger a search for the term "railgun". This is correct behavior from the bot...

But is it more intuitive when writing config for a detailed helptext for `%search` to expand to "Cadence search" without the space? I think so - If I write "Perform a search of the Cadence database (as by %search)", I prefer the result to be "Perform a search of the Cadence database (as by Cadence search)" rather than "Perform a search of the Cadence database (as by Cadence search )"

Thoughts?

za419 replied:
@JunhaoSLi @kenellorando

JunhaoSLi replied:
lgtm

za419 committed:
Talk myself into trimming whitespace from commands

If the spaces are desired, it's easy to add them in config. But if you
don't want them, it's impossible to remove them in config without a
serious redesign of the system as it stands. Therefore, we should trim
the spaces.
